### ISSUE TITLE
Test failures on Perl 5.19.10 with Perl::Critic 1.123

### ISSUE THREAD

**Issue by mjegh** [Perl-Critic/Perl-Critic#606]
```
not ok 9 - BuiltinFunctions::ProhibitBooleanGrep - line 68 - Complex booleans # TODO need to detect assignment
#   Failed (TODO) test 'BuiltinFunctions::ProhibitBooleanGrep - line 68 - Complex booleans'
#   at t/20_policies.t line 20.
# Expected 1 violations, got 0. 
ok 10 - BuiltinFunctions::ProhibitBooleanGrep - line 77 - Complex booleans
ok 11 - BuiltinFunctions::ProhibitBooleanGrep - line 85 - Complex booleans
not ok 12 - BuiltinFunctions::ProhibitBooleanGrep - line 93 - Complex booleans # TODO detect end of statement
#   Failed (TODO) test 'BuiltinFunctions::ProhibitBooleanGrep - line 93 - Complex booleans'
#   at t/20_policies.t line 20.
# Expected 1 violations, got 0. 


not ok 68 - BuiltinFunctions::ProhibitUselessTopic - line 45 - Function calls with $_ but in ways that should not be flagged.
not ok 69 - BuiltinFunctions::ProhibitUselessTopic - line 67 - GH \#600 # TODO User reported false positives
#   Failed (TODO) test 'BuiltinFunctions::ProhibitUselessTopic - line 67 - GH \#600'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 2. Found violations follow...
# Found violation: Useless use of $_ at line 2, column 15.  ($self->zilla->log($_))
# Found violation: Useless use of $_ at line 3, column 5.  (lc, $_)

not ok 73 - BuiltinFunctions::ProhibitVoidGrep - line 31 - Comma operator # TODO not handled properly

#   Failed (TODO) test 'BuiltinFunctions::ProhibitVoidGrep - line 31 - Comma operator'
#   at t/20_policies.t line 20.
# Expected 1 violations, got 0. 

not ok 246 - ControlStructures::ProhibitPostfixControls - line 204 - RT \#48422: Allow flow control method calls # TODO exemption for method calls not implimented yet
#   Failed (TODO) test 'ControlStructures::ProhibitPostfixControls - line 204 - RT \#48422: Allow flow control method calls'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 2. Found violations follow...
# Found violation: Postfix control "if" used at line 2, column 42.  (Exception::Class->throw('an expression') if $error;)
# Found violation: Postfix control "unless" used at line 3, column 39.  (Exception::Class->throw($arg1, $arg2) unless not $error;)

not ok 314 - ErrorHandling::RequireCarping - line 86 - Don't complain about obvious uses of references because they're likely being used as exception objects. # TODO not yet implemented

#   Failed (TODO) test 'ErrorHandling::RequireCarping - line 86 - Don't complain about obvious uses of references because they're likely being used as exception objects.'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 6. Found violations follow...
# Found violation: "die" used instead of "croak" at line 2, column 1.  (die \$frobnication_exception;)
# Found violation: "die" used instead of "croak" at line 3, column 1.  (die \@accumulated_warnings;)
# Found violation: "die" used instead of "croak" at line 4, column 1.  (die \%problem_data;)
# Found violation: "die" used instead of "croak" at line 6, column 1.  (die)
# Found violation: "die" used instead of "croak" at line 13, column 1.  (die)
# Found violation: "die" used instead of "croak" at line 22, column 1.  (die Blrfl::Exception->new('Too many croutons', $salad);)

not ok 422 - InputOutput::RequireBriefOpen - line 283 - unusual lexical syntax # TODO we do not recognize parenthesized lexical declarations
#   Failed (TODO) test 'InputOutput::RequireBriefOpen - line 283 - unusual lexical syntax'
#   at t/20_policies.t line 20.
# Expected 1 violations, got 0. 


not ok 449 - InputOutput::RequireCheckedClose - line 158 - autodie on and off # TODO need to handle autodie lexically.

#   Failed (TODO) test 'InputOutput::RequireCheckedClose - line 158 - autodie on and off'
#   at t/20_policies.t line 20.
# Expected 1 violations, got 0. 

not ok 467 - InputOutput::RequireCheckedOpen - line 160 - autodie on and off # TODO need to handle autodie lexically.

#   Failed (TODO) test 'InputOutput::RequireCheckedOpen - line 160 - autodie on and off'
#   at t/20_policies.t line 20.
# Expected 1 violations, got 0. 

not ok 498 - InputOutput::RequireCheckedSyscalls - line 290 - autodie on and off # TODO need to handle autodie lexically.

#   Failed (TODO) test 'InputOutput::RequireCheckedSyscalls - line 290 - autodie on and off'
#   at t/20_policies.t line 20.
# Expected 1 violations, got 0. 

not ok 743 - Modules::RequireVersionVar - line 64 - Readonly VERSION
not ok 744 - Modules::RequireVersionVar - line 72 - Readonly::Scalar VERSION

not ok 751 - NamingConventions::Capitalization - line 53 - Special case: main

not ok 1852 - NamingConventions::ProhibitAmbiguousNames - line 17 - Ambiguous word in compound name. # TODO False negative: Ambiguous words in compound names should be forbidden

#   Failed (TODO) test 'NamingConventions::ProhibitAmbiguousNames - line 17 - Ambiguous word in compound name.'
#   at t/20_policies.t line 20.
# Expected 2 violations, got 0. 

not ok 1854 - NamingConventions::ProhibitAmbiguousNames - line 39 - Ambiguous name on rhs. # TODO False positive: Need to distinguish rhs in variable statements

#   Failed (TODO) test 'NamingConventions::ProhibitAmbiguousNames - line 39 - Ambiguous name on rhs.'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 1. Found violations follow...
# Found violation: Ambiguously named variable "left" at line 2, column 1.  (my ($foo) = ($left);)

not ok 1954 - RegularExpressions::ProhibitUnusedCapture - line 234 - anon sub failures # TODO PPI v1.118 doesn't recognize anonymous subroutines

#   Failed (TODO) test 'RegularExpressions::ProhibitUnusedCapture - line 234 - anon sub failures'
#   at t/20_policies.t line 20.
# Expected 1 violations, got 0. 

not ok 1959 - RegularExpressions::ProhibitUnusedCapture - line 286 - lvalue sub assigment failure # TODO lvalue subs are too complex to support

#   Failed (TODO) test 'RegularExpressions::ProhibitUnusedCapture - line 286 - lvalue sub assigment failure'
#   at t/20_policies.t line 20.
# Expected 1 violations, got 0. 

not ok 2063 - Subroutines::ProhibitReturnSort - line 39 - when used in conjunction with wantarray() # TODO False positive: used when when wantarray() has been consulted.
#   Failed (TODO) test 'Subroutines::ProhibitReturnSort - line 39 - when used in conjunction with wantarray()'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 1. Found violations follow...
# Found violation: "return" statement followed by "sort" at line 4, column 9.  (return sort @list;)

not ok 2110 - Subroutines::RequireArgUnpacking - line 106 - failure in an anonymous sub # TODO PPI v1.118 doesn't recognize anonymous subroutines

#   Failed (TODO) test 'Subroutines::RequireArgUnpacking - line 106 - failure in an anonymous sub'
#   at t/20_policies.t line 20.
# Expected 1 violations, got 0. 

not ok 2128 - Subroutines::RequireFinalReturn - line 23 - ternary returns # TODO We are not yet detecting ternaries

#   Failed (TODO) test 'Subroutines::RequireFinalReturn - line 23 - ternary returns'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 1. Found violations follow...
# Found violation: Subroutine "foo" does not end with "return" at line 2, column 1.  (sub foo { 1 ? return : 2 ? return : return; })

not ok 2261 - ValuesAndExpressions::ProhibitCommaSeparatedStatements - line 255 - RT \#27654 - NKH example 3 # TODO PPI parses this code as blocks and not hash constructors.

#   Failed (TODO) test 'ValuesAndExpressions::ProhibitCommaSeparatedStatements - line 255 - RT \#27654 - NKH example 3'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 1. Found violations follow...
# Found violation: Comma used to separate statements at line 4, column 5.  (%{$a_hash_ref}, %{$another_hash_ref})

not ok 2456 - ValuesAndExpressions::ProhibitMixedBooleanOperators - line 71 - With a postfix control (GH \#496) # TODO need to treat left and right sides separately

#   Failed (TODO) test 'ValuesAndExpressions::ProhibitMixedBooleanOperators - line 71 - With a postfix control (GH \#496)'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 1. Found violations follow...
# Found violation: Mixed high and low-precedence booleans at line 2, column 1.  ($value ||= 1 if 1 and 1;)

not ok 2511 - ValuesAndExpressions::RequireInterpolationOfMetachars - line 220 - Email addresses with embedded violations. # TODO Policy is not smart enough to handle this yet.

#   Failed (TODO) test 'ValuesAndExpressions::RequireInterpolationOfMetachars - line 220 - Email addresses with embedded violations.'
#   at t/20_policies.t line 20.
# Expected 2 violations, got 0. 

not ok 2594 - Variables::ProhibitPerl4PackageNames - line 230 - $POSTMATCH

not ok 2615 - Variables::ProhibitPunctuationVars - line 207 - Quoted String Interpolation wart cases # TODO debug wart cases from String Interpolation exhaustive

#   Failed (TODO) test 'Variables::ProhibitPunctuationVars - line 207 - Quoted String Interpolation wart cases'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 1. Found violations follow...
# Found violation: Magic punctuation variable $\ used at line 3, column 8.  (print "$\";      # 28 of 59)

not ok 2645 - Variables::ProhibitReusedNames - line 273 - our with multiple packages - RT \#43754 # TODO We don't handle multiple packages in general, let alone in this policy.

#   Failed (TODO) test 'Variables::ProhibitReusedNames - line 273 - our with multiple packages - RT \#43754'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 1. Found violations follow...
# Found violation: Reused variable name in lexical scope: @ISA at line 5, column 1.  (our @ISA;)

not ok 2696 - Variables::RequireNegativeIndices - line 26 - Basic failure
not ok 2697 - Variables::RequireNegativeIndices - line 38 - Complex failures
```

This is perl 5, version 19, subversion 10 (v5.19.10) built for x86_64-linux

**Comment by petdance** [comment]
@mjegh Can you please try a recent Perl::Critic and let us know how it goes?

**Comment by mjegh** [comment]
I'm using perlbrew by the way. Switched back to 5.19.10 (as per initial report) and tried to install Perl-Critic-1.126.tar.gz (which cpan shell said was the latest) and got:

```
Running Build for T/TH/THALJEF/Perl-Critic-1.126.tar.gz
Building Perl-Critic


Generating t/Variables/RequireLocalizedPunctuationVars.run.
Done.



Generating t/NamingConventions/Capitalization.run.
Done.



Generating t/ControlStructures/ProhibitNegativeExpressionsInUnlessAndUntilConditions.run.
Done.

  THALJEF/Perl-Critic-1.126.tar.gz
  ./Build -- OK
Running Build test
t/00_modules.t ................................ ok         
t/01_config.t ................................. ok       
t/01_config_bad_perlcriticrc.t ................ ok     
t/01_policy_config.t .......................... ok     
t/02_policy.t ................................. ok     
t/03_annotations.t ............................ ok     
t/03_pragmas.t ................................ ok     
t/04_options_processor.t ...................... ok     
t/05_utils.t .................................. 
t/05_utils.t .................................. ok       
t/05_utils_perl.t ............................. ok   
t/05_utils_pod.t .............................. ok     
t/05_utils_ppi.t .............................. ok     
t/06_violation.t .............................. ok     
t/07_command.t ................................ ok     
t/07_perlcritic.t ............................. ok   
t/08_document.t ............................... ok     
t/09_theme.t .................................. ok     
t/10_user_profile.t ........................... ok     
t/11_policy_factory.t ......................... ok     
t/12_policy_listing.t ......................... ok       
t/12_theme_listing.t .......................... ok   
t/13_bundled_policies.t ....................... ok   
t/14_policy_parameter_behavior_boolean.t ...... ok   
t/14_policy_parameter_behavior_enumeration.t .. ok     
t/14_policy_parameter_behavior_integer.t ...... ok     
t/14_policy_parameter_behavior_list_string.t .. ok     
t/14_policy_parameter_behavior_string.t ....... ok   
t/14_policy_parameters.t ...................... ok       
t/15_statistics.t ............................. ok     
t/16_roundtrip_defaults.t ..................... ok       
t/20_policies.t ............................... 1/2706 
#   Failed test 'BuiltinFunctions::ProhibitUselessTopic - line 45 - Function calls with $_ but in ways that should not be flagged.'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 2. Found violations follow...
# Found violation: Useless use of $_ at line 15, column 17.  (my @backwards = reverse $_;)
# Found violation: Useless use of $_ at line 16, column 17.  (my @backwards = reverse($_);)
t/20_policies.t ............................... 556/2706 
#   Failed test 'Modules::RequireVersionVar - line 64 - Readonly VERSION'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 1. Found violations follow...
# Found violation: No package-scoped "$VERSION" variable found at line 1, column 1.  ()

#   Failed test 'Modules::RequireVersionVar - line 72 - Readonly::Scalar VERSION'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 1. Found violations follow...
# Found violation: No package-scoped "$VERSION" variable found at line 1, column 1.  ()

#   Failed test 'NamingConventions::Capitalization - line 53 - Special case: main'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 1. Found violations follow...
# Found violation: Package "main" does not start with a upper case letter at line 2, column 1.  (package main;)
t/20_policies.t ............................... 2423/2706 
#   Failed test 'Variables::ProhibitPerl4PackageNames - line 230 - $POSTMATCH'
#   at t/20_policies.t line 20.
# Expected 0 violations, got 2. Found violations follow...
# Found violation: "$'" uses the obsolete single quote package separator at line 2, column 8.  ($foo = $';)
# Found violation: "$'" uses the obsolete single quote package separator at line 3, column 7.  (print $';)
t/20_policies.t ............................... 2699/2706 
#   Failed test 'Variables::RequireNegativeIndices - line 26 - Basic failure'
#   at t/20_policies.t line 20.
# Expected 5 violations, got 0. 

#   Failed test 'Variables::RequireNegativeIndices - line 38 - Complex failures'
#   at t/20_policies.t line 20.
# Expected 8 violations, got 0. 
# Looks like you failed 7 tests of 2706.
t/20_policies.t ............................... Dubious, test returned 7 (wstat 1792, 0x700)
Failed 7/2706 subtests 
t/20_policy_pod_spelling.t .................... ok   
t/20_policy_prohibit_evil_modules.t ........... ok   
t/20_policy_prohibit_hard_tabs.t .............. ok     
t/20_policy_prohibit_trailing_whitespace.t .... ok   
t/20_policy_require_consistent_newlines.t ..... ok     
t/20_policy_require_tidy_code.t ............... ok   
t/92_memory_leaks.t ........................... 1/11 Unhandled type: REGEXP at /home/martin/perl5/perlbrew/perls/perl-5.19.10/lib/site_perl/5.19.10/Devel/Cycle.pm line 107.
t/92_memory_leaks.t ........................... ok     

Test Summary Report
-------------------
t/20_policies.t                             (Wstat: 1792 Tests: 2706 Failed: 7)
  Failed tests:  68, 749-750, 757, 2602, 2704-2705
  Non-zero exit status: 7
Files=38, Tests=7800, 43 wallclock secs ( 0.77 usr  0.10 sys + 38.10 cusr  2.26 csys = 41.23 CPU)
Result: FAIL
Failed 1/38 test programs. 7/7800 subtests failed.
  THALJEF/Perl-Critic-1.126.tar.gz
  ./Build test -- NOT OK
//hint// to see the cpan-testers results for installing this module, try:
  reports THALJEF/Perl-Critic-1.126.tar.gz
Failed during this command:
 THALJEF/Perl-Critic-1.126.tar.gz             : make_test NO
```

**Comment by mjegh** [comment]
I upgraded to the latest CPAN and reran - same problem.

**Comment by xsawyerx** [comment]
What's the purpose of testing against 5.19.10? It's a development release for 5.20, which is already released. Is there a problem with 5.20 or above?

**Comment by mjegh** [comment]
I wasn't testing against 5.19.10 in particular, it was just the version I had **when I originally reported the issue in 2014, 3 years ago**. My original report was for 5.19.10 and so I thought it would be moving the problem to try it on another version.

**Comment by mjegh** [comment]
Works fine on 5.20.3.

**Comment by xsawyerx** [comment]
I understand that's why this was the reported version. :)

What I'm saying is that it should be retested against the stable that came from that, specifically 5.20.x.

I believe we can now close this ticket.

**Comment by mjegh** [comment]
Closed - works in 5.20.3
